File: src/l3/dsl/ast.rs
//! AST types for the L3 hypergraph query DSL.

use serde::{Deserialize, Serialize};

/// Top-level query AST node.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Query {
    /// `MATCH (var:type) [WHERE ...] RETURN var [LIMIT n]`
    Match(NodeMatch),
    /// `MATCH HYPEREDGE e-[v1, v2, ...]- [WHERE ...] RETURN e [LIMIT n]`
    Hyperedge(HyperedgeMatch),
    /// `PATH FROM "node_id" DEPTH n [EDGE_KINDS [...]] RETURN nodes, edges`
    Path(PathQuery),
    /// `SUBGRAPH FROM "node_id" DEPTH n RETURN nodes, edges`
    Subgraph(SubgraphQuery),
}

/// Node matching query: `MATCH (n:concept) WHERE n.importance > 0.5 RETURN n LIMIT 10`
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NodeMatch {
    /// Variable name (e.g. "n")
    pub variable: Option<String>,
    /// Optional type filter (e.g. "concept")
    pub node_type: Option<String>,
    /// Optional WHERE conditions
    pub where_clause: Option<WhereCondition>,
    /// Optional LIMIT
    pub limit: Option<usize>,
}

/// Hyperedge matching query: `MATCH HYPEREDGE e-[n1, n2]- WHERE e.weight > 0.5 RETURN e`
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HyperedgeMatch {
    /// Edge variable name (e.g. "e")
    pub edge_var: Option<String>,
    /// Node variables within the hyperedge
    pub node_vars: Vec<String>,
    /// Optional WHERE conditions
    pub where_clause: Option<WhereCondition>,
    /// Optional LIMIT
    pub limit: Option<usize>,
}

/// Path traversal query: `PATH FROM "abc" DEPTH 3 EDGE_KINDS ["Related"] RETURN nodes, edges`
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PathQuery {
    /// Hex-formatted start node ID
    pub start_node: String,
    /// Maximum traversal depth
    pub max_depth: usize,
    /// Optional edge kind filter
    pub edge_kinds: Option<Vec<String>>,
}

/// Subgraph extraction query: `SUBGRAPH FROM "abc" DEPTH 2 RETURN nodes, edges`
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SubgraphQuery {
    /// Hex-formatted start node ID
    pub start_node: String,
    /// Maximum extraction depth
    pub max_depth: usize,
}

/// WHERE clause condition tree.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum WhereCondition {
    /// Property comparison: `n.importance > 0.5`
    PropertyCompare {
        property: String,
        operator: CompareOp,
        value: f32,
    },
    /// Type equality: `n.type = "concept"`
    TypeEquals(String),
    /// Keyword contains: `n.keywords CONTAINS "rust"`
    KeywordContains(String),
    /// Logical AND
    And(Box<WhereCondition>, Box<WhereCondition>),
    /// Logical OR
    Or(Box<WhereCondition>, Box<WhereCondition>),
}

/// Comparison operators for WHERE conditions.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum CompareOp {
    Gt,
    Ge,
    Lt,
    Le,
    Eq,
    Ne,
}

File: memhop/internal/core/l3/dsl/ast.go
// Copyright (c) 2026 qyiun666
// SPDX-License-Identifier: MIT OR Apache-2.0

// Package dsl implements the L3 hypergraph query DSL.
package dsl

// Query is the top-level AST node for a DSL query.
type Query struct {
	Match     *NodeMatch
	Hyperedge *HyperedgeMatch
	Path      *PathQuery
	Subgraph  *SubgraphQuery
}

// NodeMatch represents: MATCH (n:concept) WHERE ... LIMIT 10
type NodeMatch struct {
	Variable    string
	NodeType    string
	WhereClause *WhereCondition
	Limit       int
}

// HyperedgeMatch represents: MATCH HYPEREDGE e-[n1, n2]- WHERE ... LIMIT 10
type HyperedgeMatch struct {
	EdgeVar     string
	NodeVars    []string
	WhereClause *WhereCondition
	Limit       int
}

// PathQuery represents: PATH FROM "node_id" DEPTH 3 EDGE_KINDS [...]
type PathQuery struct {
	StartNode string
	MaxDepth  int
	EdgeKinds []string
}

// SubgraphQuery represents: SUBGRAPH FROM "node_id" DEPTH 2
type SubgraphQuery struct {
	StartNode string
	MaxDepth  int
}

// WhereCondition is a WHERE clause condition tree.
type WhereCondition struct {
	PropertyCompare *PropertyCompareCondition
	TypeEquals      *string
	KeywordContains *string
	And             *BinaryCondition
	Or              *BinaryCondition
}

// PropertyCompareCondition represents: n.importance > 0.5
type PropertyCompareCondition struct {
	Property string
	Operator CompareOp
	Value    float32
}

// BinaryCondition represents AND/OR of two conditions.
type BinaryCondition struct {
	Left  *WhereCondition
	Right *WhereCondition
}

// CompareOp is a comparison operator.
type CompareOp int

const (
	OpGt CompareOp = iota
	OpGe
	OpLt
	OpLe
	OpEq
	OpNe
)
